snapshots/btrfs/btrfs.go

```diff
@@ -194,47 +186,32 @@ func (b *snapshotter) View(ctx context.Context, key, parent string, opts ...snap
 	return b.makeSnapshot(ctx, snapshots.KindView, key, parent, opts)
 }
 
-func (b *snapshotter) makeSnapshot(ctx context.Context, kind snapshots.Kind, key, parent string, opts []snapshots.Opt) ([]mount.Mount, error) {
-	ctx, t, err := b.ms.TransactionContext(ctx, true)
-	if err != nil {
-		return nil, err
-	}
-	defer func() {
-		if err != nil && t != nil {
-			if rerr := t.Rollback(); rerr != nil {
-				log.G(ctx).WithError(rerr).Warn("failed to rollback transaction")
-			}
-		}
-	}()
+func (b *snapshotter) makeSnapshot(ctx context.Context, kind snapshots.Kind, key, parent string, opts []snapshots.Opt) (_ []mount.Mount, err error) {
+	var (
+		target string
+		s      storage.Snapshot
+	)
 
-	s, err := storage.CreateSnapshot(ctx, kind, key, parent, opts...)
-	if err != nil {
-		return nil, err
-	}
+	err = b.ms.WithTransaction(ctx, true, func(ctx context.Context) error {
+		s, err = storage.CreateSnapshot(ctx, kind, key, parent, opts...)
+		if err != nil {
+			return err
+		}
 
-	target := filepath.Join(b.root, strings.ToLower(s.Kind.String()), s.ID)
+		target = filepath.Join(b.root, strings.ToLower(s.Kind.String()), s.ID)
 
-	if len(s.ParentIDs) == 0 {
-		// create new subvolume
-		// btrfs subvolume create /dir
-		if err = btrfs.SubvolCreate(target); err != nil {
-			return nil, err
+		if len(s.ParentIDs) == 0 {
+			// create new subvolume
+			// btrfs subvolume create /dir
+			return btrfs.SubvolCreate(target)
 		}
-	} else {
 		parentp := filepath.Join(b.root, "snapshots", s.ParentIDs[0])
 
-		var readonly bool
-		if kind == snapshots.KindView {
-			readonly = true
-		}
-
 		// btrfs subvolume snapshot /parent /subvol
-		if err = btrfs.SubvolSnapshot(target, parentp, readonly); err != nil {
-			return nil, err
-		}
-	}
-	err = t.Commit()
-	t = nil
+		readOnly := kind == snapshots.KindView
+		return btrfs.SubvolSnapshot(target, parentp, readOnly)
+	})
+
 	if err != nil {
 		if derr := btrfs.SubvolDelete(target); derr != nil {
```

Comment:
please check target is valid. It can be empty.

Reply:
done

Reply:
If target is empty, but err is not nil.

I think it should be like

```
if err != nil {
     if target != nil {
            ...
     }
    return nil, err
}
```

Reply:
And please sign off with your real name. thanks!

Reply:
Sorry for my mistake. The new changes have been pushed